onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py

```diff
@@ -0,0 +1,565 @@
+
+        new_shape[axis] = k
+
+        for i_o in range(len(node.output)):
+            vi = self.known_vi_[node.output[i_o]]
+            vi.CopyFrom(helper.make_tensor_value_info(node.output[i_o], vi.type.tensor_type.elem_type, get_shape_from_sympy_shape(new_shape)))
+
+    def _infer_Unsqueeze(self, node):
+        self._pass_on_sympy_data(node)
+
+    def _infer_impl(self, in_mp):
+        self._set_input_model(in_mp)
+        for node in self.out_mp_.graph.node:
+            assert all([i in self.known_vi_ for i in node.input if i])
+            self._onnx_infer_single_node(node)
+            if node.op_type in self.dispatcher_:
+                self.dispatcher_[node.op_type](node)
+
+            if self.verbose_ > 2:
+                print(node.op_type + ': ' + node.name)
+            for i_o in range(len(node.output)):
+                out_shape = get_shape_from_type_proto(self.known_vi_[node.output[i_o]].type)
+                if self.verbose_ > 2:
+                    print('  ' + node.output[i_o] + ': ' + str(out_shape))
+                    if node.output[i_o] in self.sympy_data_:
+                        print('  Sympy Data: ' + str(self.sympy_data_[node.output[i_o]]))
+                if None in out_shape:
+                    if self.auto_merge_ and node.op_type in ['Add', 'Sub', 'Mul', 'Div', 'MatMul']:
+                        idx = out_shape.index(None)
+                        shapes = [self._get_shape(node, i) for i in range(2)]
+                        dim_idx = [len(s) - len(out_shape) + idx for s in shapes]
+                        assert dim_idx[0] >= 0 and dim_idx[1] >= 0
+                        dim0, dim1 = [s[i] for s, i in zip(shapes, dim_idx)]
+                        assert type(dim0) == str and type(dim1) == str
+                        if dim0 in self.computed_dims_ and dim1 in self.computed_dims_:
+                            dd = self.computed_dims_[dim0]/self.computed_dims_[dim1]
+                            dim0 = dd.args[0]
+                            dim1 = dd.args[1].args[0]
+                            assert type(dim0) == sympy.Symbol
+                            assert type(dim1) == sympy.Symbol
+                            self._add_suggested_merge(str(dim1), str(dim0))
+                        else:
+                            assert not dim0 in self.computed_dims_ or type(self.computed_dims_[dim0]) == sympy.Symbol
+                            assert not dim1 in self.computed_dims_ or type(self.computed_dims_[dim1]) == sympy.Symbol
+                            self._add_suggested_merge(str(dim1), str(dim0))
+                        self.run_ = True
+                    else:
+                        self.run_ = False
+
+                    if self.verbose_ > 0:
+                        print('Stopping at incomplete shape inference at ' + node.op_type + ': ' + node.name)
+                        for o in node.output:
```

Comment:
This should always happen otherwise failure leads to silent exit which won't be noticed unless you know the internals of the script and know that the printing of 'Done!' equates to successful completion.

Reply:
Sounds reasonable. Will change to always output.

---
In reply to: [312420519](https://github.com/microsoft/onnxruntime/pull/881#discussion_r312420519) [](ancestors = 312420519)